### ISSUE TITLE
Cleanup completed pipeline and tasks runs through dashboard

### ISSUE THREAD

**Issue by jtestard** [tektoncd/dashboard#1752]
# Is your feature request related to a problem ?

My dashboard is growing very large (lots of completed tasks) and I have no way to clean it up.

# Describe the solution you'd like

I'd like a way to delete all completed tasks runs and pipeline runs.

# Describe alternatives you've considered

I know there must be a formula to do this through `kubectl` but I haven't spent the time to construct it. If the workaround through CLI could be added in the comment, it would help a lot, thanks :).

**Comment by eddycharly** [comment]
Hi @jtestard !
It’s not really a dashboard specific issue. I think we have an example of cleaning up script in the plumbing repo. Let me check...

**Comment by jtestard** [comment]
My point is if someone uses only the tekton dashboard, this is one of the
tasks that quickly becomes necessary

**Comment by eddycharly** [comment]
Couldn’t find it, maybe @afrittoli knows the answer better ;)
At some point it will be an issue for every component. No cleanup means longer reconciliation time at startup etc... but I agree that it manifests quickly with the dashboard.

**Comment by jtestard** [comment]
In general, at my company we are using tekton as early adopters because we are excited by the technology, but we are facing a number of problems that make tekton not quite "production-ready" yet. Happy to chat more about this. This issue is one of them.

**Comment by AlanGreene** [comment]
Similar requests have come up before. The `tkn` CLI provides some help for this. For example:

`tkn pr delete -p <pipeline-name> --keep 2`

would delete all PipelineRuns of the specified Pipeline, keeping only the 2 most recent ones.

This is the approach used in our plumbing repo, see https://github.com/tektoncd/plumbing/blob/698db03c9f9888c46a7219d9a40c6b2f24763d51/tekton/resources/cd/cleanup-template.yaml#L60

There's also an issue to add support for field selectors which might allow to query based on status or similar, however this is currently blocked by lack of support in Kubernetes, see https://github.com/tektoncd/cli/pull/995#issuecomment-660182582 for more.

One option we've considered for the dashboard is the ability to select multiple runs in the dashboard UI and batch delete them, similar to what we currently support for Secrets.

**Comment by eddycharly** [comment]
IMO, tekton is un opinionated and you should not expect tekton to do things other than what it is supposed to do.

Some will use the dashboard while others won’t, some will want to keep everything around while some need some sort of garbage collection. Tekton won’t do anything on your back, maybe it should be more clear so that users don’t get surprised.

We often see similar requests though and it would be nice to have a place to provide guidance and tips because answering similar requests on a case by case basis is not effective.

**Comment by eddycharly** [comment]
IMO, tekton is un opinionated and you should not expect tekton to do things other than what it is supposed to do.

Some will use the dashboard while others won’t, some will want to keep everything around while some need some sort of garbage collection. Tekton won’t do anything on your back, maybe it should be more clear so that users don’t get surprised.

We often see similar requests though and it would be nice to have a place to provide guidance and tips because answering similar requests on a case by case basis is not effective.

**Comment by jtestard** [comment]
> One option we've considered for the dashboard is the ability to select multiple runs in the dashboard UI and batch delete them, similar to what we currently support for Secrets and PipelineResources.

This is interesting to us.

> IMO, tekton is un opinionated and you should not expect tekton to do things other than what it is supposed to do.

Getting out of scope for this issue, but as long as there's an ecosystem in which add-ons can thrive, we could build our own for this kind of purpose as well.

**Comment by AlanGreene** [comment]
Assigning to @steveodonovan to implement the batch delete piece of this.

**Comment by AlanGreene** [comment]
There's now an example in the experimental repo with documentation showing how to use a CronJob to clean up completed runs. This is based on the approach from plumbing that I described above.

https://github.com/tektoncd/experimental/tree/721e8c4ec2848a609e436b2c392712b64d1cc792/pipeline/cleanup
